src/main/kotlin/org/jitsi/nlj/transform/node/Node.kt

```diff
@@ -322,34 +322,61 @@ sealed class StatsKeepingNode(name: String) : Node(name) {
     }
 }
 
+/**
+ * A node which will always forward the [PacketInfo] it is given.
+ * NOTE that the [PacketInfo] instance may be modified, but only the original
+ * [PacketInfo] instance will be forwarded.
+ */
+abstract class NeverDiscardNode(name: String) : StatsKeepingNode(name) {
+    protected abstract fun handlePacket(packetInfo: PacketInfo)
+
+    final override fun doProcessPacket(packetInfo: PacketInfo) {
+        handlePacket(packetInfo)
+        doneProcessing(packetInfo)
+        next(packetInfo)
+    }
+}
+
 /**
  * A [Node] which transforms a single packet, possibly dropping it (by returning null).
+ * If null is returned, the [PacketInfo] instance given to [transform] will be
+ * discarded.
  */
 abstract class TransformerNode(name: String) : StatsKeepingNode(name) {
-
     protected abstract fun transform(packetInfo: PacketInfo): PacketInfo?
 
     override fun doProcessPacket(packetInfo: PacketInfo) {
         val transformedPacket = transform(packetInfo)
         doneProcessing(transformedPacket)
-        if (transformedPacket != null) {
-            next(transformedPacket)
+        when (transformedPacket) {
+            null -> packetDiscarded(packetInfo)
+            else -> next(transformedPacket)
```

Comment:
Is there any way to ensure that you've cleaned up all the `packetDiscarded` calls in subclasses of this?  Maybe have this call `super.packetDiscarded` and make this class's `packetDiscarded` assert or something?

Reply:
I don't think I'm following you.  This class is effectively calling `super.packetDiscarded` here (just the `super` is omitted), and the `packetDiscarded` method isn't `open`.

Reply:
What I meant was that the derived classes of this class used to call `packetDiscarded`, and now don't -- and if you happened to miss some call to `packetDiscarded` you'd double-discard the packet and bad things could happen.

If we made `packetDiscarded` open and changed this class's implementation of it to assert, we could potentially protect against this.  Probably not worth it, though.

Reply:
Ok, yeah I can do that.